src/dialer.js

```diff
@@ -0,0 +1,185 @@
+  constructor (perPeerLimit, dialTimeout) {
+    log('create: %s peer limit, %s dial timeout', perPeerLimit, dialTimeout)
+    this.perPeerLimit = perPeerLimit
+    this.dialTimeout = dialTimeout
+    this.queues = new Map()
+  }
+
+  /**
+   * Dial a list of multiaddrs on the given transport.
+   *
+   * @param {PeerId} peer
+   * @param {SwarmTransport} transport
+   * @param {Array<Multiaddr>} addrs
+   * @param {function(Error, Connection)} callback
+   * @returns {void}
+   */
+  dialMany (peer, transport, addrs, callback) {
+    log('dialMany:start')
+    // we use a token to track if we want to cancel following dials
+    const token = {cancel: false}
+    map(addrs, (m, cb) => {
+      this.dialSingle(peer, transport, m, token, cb)
+    }, (err, results) => {
+      if (err) {
+        return callback(err)
+      }
+
+      const success = results.filter((res) => res.conn)
+      if (success.length > 0) {
+        log('dialMany:success')
+        return callback(null, success[0].conn)
+      }
+
+      log('dialMany:error')
+      const error = new Error('Failed to dial any provided address')
+      error.errors = results
+        .filter((res) => res.error)
+        .map((res) => res.error)
+      return callback(error)
+    })
+  }
+
+  /**
+   * Dial a single multiaddr on the given transport.
+   *
+   * @param {PeerId} peer
+   * @param {SwarmTransport} transport
+   * @param {Multiaddr} addr
+   * @param {CancelToken} token
+   * @param {function(Error, Connection)} callback
+   * @returns {void}
+   */
+  dialSingle (peer, transport, addr, token, callback) {
+    const ps = peer.toB58String()
+    log('dialSingle: %s:%s', ps, addr.toString())
+    let q
+    if (this.queues.has(ps)) {
+      q = this.queues.get(ps)
+    } else {
+      q = new DialQueue(this.perPeerLimit, this.dialTimeout)
+      this.queues.set(ps, q)
+    }
+
+    q.push(transport, addr, token, callback)
+  }
+}
+
+/**
+ * Queue up the amount of dials to a given peer.
+ */
+class DialQueue {
+  /**
+   * Create a new dial queue.
+   *
+   * @param {number} limit
+   * @param {number} dialTimeout
+   */
+  constructor (limit, dialTimeout) {
+    this.dialTimeout = dialTimeout
+
+    this.queue = queue((task, cb) => {
+      this._doWork(task.transport, task.addr, task.token, cb)
+    }, limit)
+  }
+
+  /**
+   * The actual work done by the queue.
+   *
+   * @param {SwarmTransport} transport
+   * @param {Multiaddr} addr
+   * @param {CancelToken} token
+   * @param {function(Error, Connection)} callback
+   * @returns {void}
+   * @private
+   */
+  _doWork (transport, addr, token, callback) {
+    log('dialQueue:work')
+    this._dialWithTimeout(
+      transport,
+      addr,
+      (err, conn) => {
+        if (err) {
+          log('dialQueue:work:error')
+          return callback(null, {error: err})
+        }
+
+        if (token.cancel) {
+          log('dialQueue:work:cancel')
+          // clean up already done dials
+          pull(pull.empty(), conn)
```

Comment:
Just to make sure we don't loose track of this:

- https://github.com/libp2p/js-libp2p-swarm/pull/200#discussion_r108214454
- https://github.com/libp2p/js-libp2p-swarm/pull/200#issuecomment-289577321